src/asset_manager/mod.rs

```diff
@@ -1,4 +1,221 @@
+//!
+//! For how to implement an asset yourself, see the `Asset` trait.
+//!
+//! If you just want to load it, look at `AssetLoader` / `AssetManager`.
 
-mod asset_manager;
-pub use self::asset_manager::*;
+pub mod formats;
+
+mod asset;
+mod common;
+mod io;
+
+pub use self::asset::{Asset, AssetFormat, AssetStore, AssetStoreError};
+pub use self::common::{DefaultStore, DirectoryStore, ZipStore};
+pub use self::io::{Import, Error as ImportError};
+
+use std::fmt::{Debug, Display, Error as FormatError, Formatter};
+use std::marker::Send;
+
+use futures::{Async, Future};
+use futures_cpupool::{CpuPool, CpuFuture};
+
+use engine::Context;
+
+/// A parallel asset loader.
+/// It contains a `CpuPool` and
+/// creates `Future`s from that.
+///
+/// Only the `Asset::Data` will be
+/// loaded in a separate thread, the
+/// asset itself will be created on the main
+/// thread, because often it requires thread-unsafe
+/// actions (accessing OpenGL to create some buffer).
+///
+/// The asset loader needs three things in order to
+/// load an asset:
+///
+/// * An asset store (`AssetStore`): Responsible for providing bytes for a name and a format
+/// * The asset name: A simple `&str`
+/// * The asset format (`AssetFormat`): Has to implement `Import` (create the `Asset::Data` structure) and provide the typical file extension
+///
+/// # Examples
+///
+/// ```ignore
+/// use amethyst::asset_manager::AssetLoader;
+///
+/// let loader = AssetLoader::new();
+/// let asset_future = loader.load(my_store, "my_asset", MyFormat);
+/// ```
+pub struct AssetLoader {
+    cpupool: CpuPool,
+}
+
+/// An asset future which means
+/// "This will be available in the future".
+///
+/// #### Why isn't it available immediately?
+///
+/// The reason is that assets are loaded in
+/// separate threads, without blocking the
+/// main thread. Once they're finished,
+/// you can access it.
+///
+/// As soon as you need the asset, you call
+/// `finish` on this future which will return
+/// the loaded data and create an asset from that
+/// data. If it hasn't yet finished, it will block
+/// the calling thread.
+pub struct AssetFuture<T: Asset> {
+    inner: CpuFuture<T::Data, AssetError<T>>,
+}
+
+/// The error that can occurr when trying
+/// to import asset data from a store.
+#[derive(Debug)]
+pub enum AssetError<T: Asset> {
+    /// Occurs if the `AssetStore` could not load
+    /// the asset. See `AssetStoreInfo` for details.
+    StoreError(AssetStoreError),
+    /// Raised if the data is in an invalid format
+    /// or there was an io error.
+    ImportError(ImportError),
+    /// Importing the data didn't work
+    /// out. Note that this does not necessarily mean
+    /// that the data is invalid.
+    DataError(T::Error),
+}
+
+impl AssetLoader {
+    /// Creates a new asset loader with a cpu pool
+    /// using the number of cpu cores for the number of threads.
+    pub fn new() -> Self {
+        AssetLoader { cpupool: CpuPool::new_num_cpus() }
```

Comment:
For that to work `specs` has to use `futures` though. I'll try to get that working, but it will take me quite some time because I don't know how it's working exactly atm (with this `Pulse` and the channel)